{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Implementation of a DanQ-like network on sequences landing in peaks of H3K27ac ChIP-Seq data from three different cell types (neuron, glia and microglia) sorted from the dorsolateral prefrontal cortex of postmortem healthy brains\n",
    "\n",
    "Package dependencies: intervaltree, numpy, pandas, ucscgenome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing packages, initializing input file names and other variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import intervaltree\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import ucscgenome\n",
    "\n",
    "\n",
    "genome_object = ucscgenome.Genome('hg19', cache_dir='/home/taeyoonp/multiple_alignment-python/seqfiles/', use_web=False)\n",
    "peak_file_dir = '/projects/pfenninggroup/jemmie/3expHg/rawData/bam/case_control_peaks_stringent/'\n",
    "\n",
    "label_names = ['G', 'M', 'N']\n",
    "\n",
    "narrow_peak_file_paths ={\n",
    "                           'G':os.path.join(peak_file_dir,'controls_Glia_peaks_peaks.narrowPeak'),\n",
    "                           'M':os.path.join(peak_file_dir,'controls_Microglia_peaks_peaks.narrowPeak'),\n",
    "                           'N':os.path.join(peak_file_dir,'controls_Neuron_peaks_peaks.narrowPeak')\n",
    "                        }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Function to parse a narrow peak bed file and retrieve coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_narrow_peak_bed_file(narrow_peak_file, reference_genome_name):\n",
    "    # headers specific to narrow peak format (order matters)\n",
    "    narrow_peak_headers = [\"chromosome_name\", \"chromosome_start_index\", \"chromosome_end_index\",\n",
    "                           \"region_name\", \"score\", \"strand\", \"signal_value\", \"p_value\", \"q_value\", \"peak\"]\n",
    "\n",
    "    with open(narrow_peak_file, 'r') as narrow_peak_bed_csv:\n",
    "        narrow_peak_df = pd.read_csv(narrow_peak_bed_csv, sep='\\t', header=None, names=narrow_peak_headers)\n",
    "        narrow_peak_df['reference_genome'] = reference_genome_name\n",
    "    return narrow_peak_df.to_dict(orient=\"records\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Function to get the overlap length between two sets of coordinates on a chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calcOverlapLength(first_start, first_end, second_start, second_end):\n",
    "    if first_start > second_end:\n",
    "        return 0\n",
    "    elif second_start > first_end:\n",
    "        return 0\n",
    "    return min(first_end, second_end) - max(first_start, second_start) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Function to check whether a set of coordinates lies at least half in peak regions (uses interval tree API)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def are_coordinates_half_in_peaks(trees_by_chromosome, chromosome, start, stop):\n",
    "    if not chromosome in trees_by_chromosome:\n",
    "        return 0\n",
    "    \n",
    "    half_length = (stop-start)/2\n",
    "    \n",
    "    chromosome_peak_tree = trees_by_chromosome[chromosome]\n",
    "    \n",
    "    overlapPeaks = chromosome_peak_tree[start:stop]\n",
    "    \n",
    "    overlapLength = 0\n",
    "    \n",
    "    for overlapPeak in overlapPeaks:\n",
    "        currOverlap = calcOverlapLength(overlapPeak.begin, overlapPeak.end, start, stop)\n",
    "        overlapLength+=currOverlap\n",
    "        \n",
    "    return 1 if overlapLength >= half_length else 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Constructing interval tree from peak data for quickly querying and calculating overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def construct_interval_tree(peak_data):\n",
    "    trees_by_chromosome = dict()\n",
    "    for peak in peak_data:\n",
    "        chromosome = peak['chromosome_name']\n",
    "        start_index = peak['chromosome_start_index']\n",
    "        end_index = peak['chromosome_end_index']\n",
    "        \n",
    "        if chromosome not in trees_by_chromosome:\n",
    "            trees_by_chromosome[chromosome] = intervaltree.IntervalTree()\n",
    "        trees_by_chromosome[chromosome].addi(start_index, end_index)\n",
    "    return trees_by_chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_one_hot_encoded_sequence(sequence):\n",
    "    ONE_HOT_DIMENSION = (len(sequence), 4)\n",
    "    DNA_ALPHABET = {\"A\":0, \"G\":1, \"C\":2, \"T\":3}    \n",
    "    one_hot_encoded_sequence = np.zeros(ONE_HOT_DIMENSION, dtype=np.int)\n",
    "    for i, nucleotide in enumerate(sequence):\n",
    "        if nucleotide.upper() in DNA_ALPHABET:\n",
    "            index = DNA_ALPHABET[nucleotide.upper()]\n",
    "            one_hot_encoded_sequence[i][index] = 1\n",
    "    return one_hot_encoded_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the data from input narrow peak files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "data_dict = {}\n",
    "\n",
    "for label in label_names:\n",
    "    data_dict[label] = parse_narrow_peak_bed_file(narrow_peak_file_paths[label], genome_object.genome_file)\n",
    "\n",
    "peak_coordinate_trees = {}\n",
    "\n",
    "for label in label_names:\n",
    "    peak_coordinate_trees[label] = construct_interval_tree(data_dict[label])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def construct_keras_input_for_chromosome(chromosome, bin_size, flank_size, genome_object, peak_coordinate_trees, label_names):    \n",
    "    X = []\n",
    "    Y = []\n",
    "    metadata = []\n",
    "    chromosome_size = genome_object.sequence_sizes()[chromosome]  \n",
    "    for i in xrange(0, chromosome_size, bin_size):\n",
    "        response_variables = {}\n",
    "        is_an_example = False\n",
    "        \n",
    "        responses = [are_coordinates_half_in_peaks(peak_coordinate_trees[label], chromosome, i, i+bin_size) for label in label_names]        \n",
    "        is_an_example = any(responses)\n",
    "\n",
    "        if is_an_example:\n",
    "            input_sequence = get_one_hot_encoded_sequence(genome_object[chromosome][i-flank_size:i+bin_size+flank_size])\n",
    "            X.append(input_sequence)\n",
    "            Y.append(responses)\n",
    "            metadata.append([chromosome, i, i+bin_size])\n",
    "    \n",
    "    X = np.stack(X, axis=0)\n",
    "    Y = np.stack(Y, axis=0)\n",
    "    metadata = np.stack(metadata,axis=0)\n",
    "    \n",
    "    return X, Y, metadata                                                                      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def construct_keras_input_for_chromosomes(chromosomes, bin_size, flank_size, genome_object, peak_coordinate_trees, label_names):    \n",
    "    X = []\n",
    "    Y = []\n",
    "    metadata = []\n",
    "\n",
    "    for chromosome in chromosomes:\n",
    "        chromosome_X_data, chromosome_Y_data, chromosome_metadata = construct_keras_input_for_chromosome(chromosome, bin_size, flank_size, genome_object, peak_coordinate_trees, label_names)\n",
    "        X.append(chromosome_X_data)\n",
    "        Y.append(chromosome_Y_data)\n",
    "        metadata.append(chromosome_metadata)\n",
    "\n",
    "    X = np.concatenate(X)\n",
    "    Y = np.concatenate(Y)\n",
    "    metadata = np.concatenate(metadata)\n",
    "    return X, Y, metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_chromosomes = [\n",
    "                        'chr1',\n",
    "                        'chr10',\n",
    "                        'chr11',\n",
    "                        'chr12',\n",
    "                        'chr13',\n",
    "                        'chr14',\n",
    "                        'chr15',\n",
    "                        'chr16',\n",
    "                        'chr17',\n",
    "                        'chr18',\n",
    "                        'chr19',\n",
    "                        'chr2',\n",
    "                        'chr20',\n",
    "                        'chr21',\n",
    "                        'chr22',\n",
    "                        'chr3',\n",
    "                        'chr5',\n",
    "                        'chr6',\n",
    "                        'chr7',\n",
    "                        'chrX',\n",
    "                        'chrY'\n",
    "                    ]\n",
    "\n",
    "test_chromosomes =  [\n",
    "                        'chr8',\n",
    "                        'chr9',\n",
    "                    ]\n",
    "\n",
    "valid_chromosomes = [\n",
    "                        'chr4',\n",
    "                    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bin_size = 200\n",
    "flank_size = 400\n",
    "\n",
    "X_train, Y_train, metadata_train = construct_keras_input_for_chromosomes(train_chromosomes, bin_size, flank_size, genome_object, peak_coordinate_trees, label_names)\n",
    "X_test, Y_test, metadata_test = construct_keras_input_for_chromosomes(test_chromosomes, bin_size, flank_size, genome_object, peak_coordinate_trees, label_names)\n",
    "X_valid, Y_valid, metadata_valid = construct_keras_input_for_chromosomes(valid_chromosomes, bin_size, flank_size, genome_object, peak_coordinate_trees, label_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "np.save('/home/eramamur/X_train.npy', X_train)\n",
    "np.save('/home/eramamur/Y_train.npy', Y_train)\n",
    "np.save('/home/eramamur/metadata_train.npy', metadata_train)\n",
    "np.save('/home/eramamur/X_test.npy', X_test)\n",
    "np.save('/home/eramamur/Y_test.npy', Y_test)\n",
    "np.save('/home/eramamur/metadata_test.npy', metadata_test)\n",
    "np.save('/home/eramamur/X_valid.npy', X_valid)\n",
    "np.save('/home/eramamur/Y_valid.npy', Y_valid)\n",
    "np.save('/home/eramamur/metadata_valid.npy', metadata_valid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "compiling model\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_4 (Conv1D)            (None, 975, 320)          33600     \n",
      "_________________________________________________________________\n",
      "max_pooling1d_4 (MaxPooling1 (None, 75, 320)           0         \n",
      "_________________________________________________________________\n",
      "dropout_7 (Dropout)          (None, 75, 320)           0         \n",
      "_________________________________________________________________\n",
      "bidirectional_7 (Bidirection (None, 75, 640)           1640960   \n",
      "_________________________________________________________________\n",
      "bidirectional_8 (Bidirection (None, 75, 640)           2460160   \n",
      "_________________________________________________________________\n",
      "dropout_8 (Dropout)          (None, 75, 640)           0         \n",
      "_________________________________________________________________\n",
      "flatten_4 (Flatten)          (None, 48000)             0         \n",
      "_________________________________________________________________\n",
      "dense_7 (Dense)              (None, 925)               44400925  \n",
      "_________________________________________________________________\n",
      "activation_7 (Activation)    (None, 925)               0         \n",
      "_________________________________________________________________\n",
      "dense_8 (Dense)              (None, 3)                 2778      \n",
      "_________________________________________________________________\n",
      "activation_8 (Activation)    (None, 3)                 0         \n",
      "=================================================================\n",
      "Total params: 48,538,423.0\n",
      "Trainable params: 48,538,423.0\n",
      "Non-trainable params: 0.0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "from keras.models import Sequential\n",
    "\n",
    "\n",
    "from keras.layers.convolutional import Conv1D\n",
    "from keras.layers.core import Dense, Dropout, Activation, Flatten\n",
    "from keras.layers.pooling import MaxPooling1D\n",
    "from keras.layers.wrappers import Bidirectional\n",
    "from keras.layers.recurrent import LSTM\n",
    "\n",
    "from keras.utils import plot_model\n",
    "\n",
    "\n",
    "\n",
    "model = Sequential()\n",
    "\n",
    "model.add(Conv1D(input_shape = (1000, 4),\n",
    "                    padding=\"valid\",\n",
    "                    strides=1,\n",
    "                    activation=\"relu\",\n",
    "                    kernel_size=26,\n",
    "                    filters=320))                 \n",
    "                 \n",
    "          \n",
    "model.add(MaxPooling1D(pool_size=13, strides=13))\n",
    "\n",
    "model.add(Dropout(0.2))\n",
    "      \n",
    "model.add(Bidirectional(LSTM(320, return_sequences=True)))\n",
    "model.add(Bidirectional(LSTM(320, return_sequences=True)))          \n",
    "          \n",
    "model.add(Dropout(0.5))\n",
    "\n",
    "model.add(Flatten())\n",
    "\n",
    "model.add(Dense(input_dim=75*640, units=925))\n",
    "model.add(Activation('relu'))\n",
    "\n",
    "model.add(Dense(units=3))\n",
    "model.add(Activation('sigmoid'))\n",
    "\n",
    "print 'compiling model'\n",
    "model.compile(loss='binary_crossentropy', optimizer='rmsprop')\n",
    "\n",
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train on 1341018 samples, validate on 65406 samples\n",
      "Epoch 1/5\n",
      " 204900/1341018 [===>..........................] - ETA: 100714s - loss: 7.1682"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-2b1a270cc6c9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mY_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshuffle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mX_valid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_valid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minitial_epoch\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/home/eramamur/bin/miniconda2/lib/python2.7/site-packages/keras/models.pyc\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, **kwargs)\u001b[0m\n\u001b[1;32m    843\u001b[0m                               \u001b[0mclass_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclass_weight\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    844\u001b[0m                               \u001b[0msample_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_weight\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 845\u001b[0;31m                               initial_epoch=initial_epoch)\n\u001b[0m\u001b[1;32m    846\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    847\u001b[0m     def evaluate(self, x, y, batch_size=32, verbose=1,\n",
      "\u001b[0;32m/home/eramamur/bin/miniconda2/lib/python2.7/site-packages/keras/engine/training.pyc\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, **kwargs)\u001b[0m\n\u001b[1;32m   1483\u001b[0m                               \u001b[0mval_f\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mval_f\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mval_ins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mval_ins\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshuffle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mshuffle\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1484\u001b[0m                               \u001b[0mcallback_metrics\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcallback_metrics\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1485\u001b[0;31m                               initial_epoch=initial_epoch)\n\u001b[0m\u001b[1;32m   1486\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1487\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mevaluate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m32\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/eramamur/bin/miniconda2/lib/python2.7/site-packages/keras/engine/training.pyc\u001b[0m in \u001b[0;36m_fit_loop\u001b[0;34m(self, f, ins, out_labels, batch_size, epochs, verbose, callbacks, val_f, val_ins, shuffle, callback_metrics, initial_epoch)\u001b[0m\n\u001b[1;32m   1138\u001b[0m                 \u001b[0mbatch_logs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'size'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch_ids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1139\u001b[0m                 \u001b[0mcallbacks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mon_batch_begin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbatch_index\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_logs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1140\u001b[0;31m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mins_batch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1141\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mouts\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1142\u001b[0m                     \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mouts\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/eramamur/bin/miniconda2/lib/python2.7/site-packages/keras/backend/theano_backend.pyc\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs)\u001b[0m\n\u001b[1;32m   1092\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1093\u001b[0m         \u001b[0;32massert\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1094\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfunction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1095\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1096\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/eramamur/bin/miniconda2/lib/python2.7/site-packages/theano/compile/function_module.pyc\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    882\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    883\u001b[0m             \u001b[0moutputs\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 884\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0moutput_subset\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    885\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_subset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0moutput_subset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    886\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/eramamur/bin/miniconda2/lib/python2.7/site-packages/theano/scan_module/scan_op.pyc\u001b[0m in \u001b[0;36mrval\u001b[0;34m(p, i, o, n, allow_gc)\u001b[0m\n\u001b[1;32m    985\u001b[0m         \u001b[0mallow_gc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mallow_gc\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mallow_gc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    986\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 987\u001b[0;31m         def rval(p=p, i=node_input_storage, o=node_output_storage, n=node,\n\u001b[0m\u001b[1;32m    988\u001b[0m                  allow_gc=allow_gc):\n\u001b[1;32m    989\u001b[0m             \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "model.fit(x=X_train, y=Y_train, batch_size=100, epochs=5, shuffle=True, verbose=1, validation_data = (X_valid, Y_valid), initial_epoch=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Below functions not needed right now. Was used earlier but interval trees are more efficient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_sorted_peak_coordinates(peak_data):\n",
    "    all_peak_coordinates = dict()\n",
    "    for peak in peak_data:\n",
    "        chromosome = peak['chromosome_name']\n",
    "        start_index = peak['chromosome_start_index']\n",
    "        end_index = peak['chromosome_end_index']\n",
    "        \n",
    "        if chromosome not in all_peak_coordinates:\n",
    "            all_peak_coordinates[chromosome] = []\n",
    "        all_peak_coordinates[chromosome].append([start_index, end_index])\n",
    "\n",
    "    for chromosome in all_peak_coordinates:\n",
    "        all_peak_coordinates[chromosome] = sorted(all_peak_coordinates[chromosome], key=lambda x: (x[0],x[1]))\n",
    "    return all_peak_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def testCalcOverlapLength():\n",
    "    peak_start = 100\n",
    "    peak_end = 700\n",
    "    coord_start = 500\n",
    "    coord_end = 700\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==200)\n",
    "    coord_start = 0\n",
    "    coord_end = 99\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==0)\n",
    "    coord_start = 0\n",
    "    coord_end = 400\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==300)\n",
    "    coord_start= 200\n",
    "    coord_end = 400\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==200)\n",
    "    coord_start = 50\n",
    "    coord_end = 740\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==600)\n",
    "    coord_start = 0\n",
    "    coord_end = 100\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==0)\n",
    "    coord_start = 700\n",
    "    coord_end = 900\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==0)\n",
    "    coord_start = 0\n",
    "    coord_end = 101\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==1)\n",
    "    coord_start = 699\n",
    "    coord_end = 899\n",
    "    assert(calcOverlapLength(peak_start, peak_end, coord_start, coord_end)==1)    \n",
    "    \n",
    "testCalcOverlapLength()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'bool' object has no attribute '__getitem__'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-c069efa34e2a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    291\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msorted_peak_coordinates\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'chr11'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 293\u001b[0;31m \u001b[0mtest_are_coordinates_half_in_peaks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-15-c069efa34e2a>\u001b[0m in \u001b[0;36mtest_are_coordinates_half_in_peaks\u001b[0;34m()\u001b[0m\n\u001b[1;32m    260\u001b[0m     \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m540470\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m     \u001b[0mstop\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m540670\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 262\u001b[0;31m     \u001b[0;32massert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mare_coordinates_half_in_peaks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted_peak_coordinates\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchromosome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    264\u001b[0m     \u001b[0mchromosome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'chr1'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'bool' object has no attribute '__getitem__'"
     ]
    }
   ],
   "source": [
    "def test_are_coordinates_half_in_peaks():    \n",
    "    \"\"\"THIS TEST WAS SUPPOSED TO WORK FOR A PREVIOUS FUNCTION VERSION\"\"\"\n",
    "    sorted_peak_coordinates = {'chr1': [[540603, 540681],\n",
    "                                      [713280, 713471],\n",
    "                                      [713888, 714341],\n",
    "                                      [714450, 714755],\n",
    "                                      [762430, 762597],\n",
    "                                      [762657, 763009],\n",
    "                                      [785042, 785218],\n",
    "                                      [785399, 785647]],\n",
    "                                'chr10': [[177214, 177361],\n",
    "                                      [9972654, 9972743],\n",
    "                                      [9973249, 9973451],\n",
    "                                      [9973703, 9974065],\n",
    "                                      [9974388, 9974520]],\n",
    "                                 'chr11': [[187562, 187823],\n",
    "                                  [188147, 188324],\n",
    "                                  [189338, 190324],\n",
    "                                  [190409, 190945],\n",
    "                                  [191958, 193127],\n",
    "                                  [193181, 193802],\n",
    "                                  [193857, 194098],\n",
    "                                  [194152, 194274],\n",
    "                                  [194361, 194833],\n",
    "                                  [195038, 196019],\n",
    "                                  [196845, 196947],\n",
    "                                  [199618, 199785],\n",
    "                                  [206704, 209406],\n",
    "                                  [209480, 209604],\n",
    "                                  [210355, 210658],\n",
    "                                  [210895, 210968],\n",
    "                                  [211227, 211305],\n",
    "                                  [211594, 211695],\n",
    "                                  [212835, 212908],\n",
    "                                  [213103, 213179],\n",
    "                                  [213230, 213393],\n",
    "                                  [214011, 214438],\n",
    "                                  [215456, 215536],\n",
    "                                  [215767, 216267],\n",
    "                                  [216336, 216460],\n",
    "                                  [217046, 217144],\n",
    "                                  [218204, 219831],\n",
    "                                  [220558, 220810],\n",
    "                                  [221072, 223807],\n",
    "                                  [224043, 224957],\n",
    "                                  [226100, 226275],\n",
    "                                  [227488, 227577],\n",
    "                                  [227932, 228052],\n",
    "                                  [235354, 235456],\n",
    "                                  [235523, 237520],\n",
    "                                  [288598, 288700],\n",
    "                                  [288781, 288918],\n",
    "                                  [355736, 355995],\n",
    "                                  [356240, 356419],\n",
    "                                  [368586, 369026],\n",
    "                                  [369099, 369294],\n",
    "                                  [370071, 372365],\n",
    "                                  [373145, 373254],\n",
    "                                  [373582, 374143],\n",
    "                                  [374239, 374614],\n",
    "                                  [375220, 376089],\n",
    "                                  [376441, 378182],\n",
    "                                  [379687, 379835],\n",
    "                                  [381818, 382050],\n",
    "                                  [382239, 382341],\n",
    "                                  [382433, 382650],\n",
    "                                  [384800, 384883],\n",
    "                                  [385698, 386675],\n",
    "                                  [386769, 388769],\n",
    "                                  [392943, 393017],\n",
    "                                  [395881, 396106],\n",
    "                                  [396202, 396522],\n",
    "                                  [396574, 397632],\n",
    "                                  [397699, 397824],\n",
    "                                  [406597, 406732],\n",
    "                                  [406792, 407089],\n",
    "                                  [447937, 449038],\n",
    "                                  [449614, 449688],\n",
    "                                  [449827, 451974],\n",
    "                                  [455313, 455907],\n",
    "                                  [456490, 457993],\n",
    "                                  [458074, 458151],\n",
    "                                  [460394, 460486],\n",
    "                                  [461374, 461459],\n",
    "                                  [461717, 461942],\n",
    "                                  [462562, 462670],\n",
    "                                  [464316, 464396],\n",
    "                                  [464448, 464702],\n",
    "                                  [467997, 470642],\n",
    "                                  [470901, 471083],\n",
    "                                  [471269, 471691],\n",
    "                                  [471813, 473174],\n",
    "                                  [473284, 473361],\n",
    "                                  [473429, 473532],\n",
    "                                  [474112, 474292],\n",
    "                                  [476399, 476550],\n",
    "                                  [476793, 477900],\n",
    "                                  [479081, 479225],\n",
    "                                  [479341, 479473],\n",
    "                                  [487532, 487610],\n",
    "                                  [487707, 487817],\n",
    "                                  [487912, 488051],\n",
    "                                  [488615, 489262],\n",
    "                                  [490030, 490196],\n",
    "                                  [490394, 490493],\n",
    "                                  [490648, 490788],\n",
    "                                  [490856, 491674],\n",
    "                                  [492660, 492743],\n",
    "                                  [492981, 493195],\n",
    "                                  [494194, 494410],\n",
    "                                  [494578, 494865],\n",
    "                                  [494928, 495881],\n",
    "                                  [495939, 496278],\n",
    "                                  [498152, 498373],\n",
    "                                  [498568, 498806],\n",
    "                                  [499076, 499238],\n",
    "                                  [499302, 499441],\n",
    "                                  [499517, 499690],\n",
    "                                  [499781, 500298],\n",
    "                                  [501083, 501174],\n",
    "                                  [501610, 501710],\n",
    "                                  [501844, 503116],\n",
    "                                  [503234, 503395],\n",
    "                                  [503624, 504822],\n",
    "                                  [506154, 507346],\n",
    "                                  [507473, 507883],\n",
    "                                  [507997, 508076],\n",
    "                                  [518468, 518776],\n",
    "                                  [519032, 519108],\n",
    "                                  [533615, 533731],\n",
    "                                  [534107, 535414],\n",
    "                                  [535629, 536469],\n",
    "                                  [536710, 537288],\n",
    "                                  [537370, 538225],\n",
    "                                  [538385, 538482],\n",
    "                                  [554688, 554809],\n",
    "                                  [554912, 554985],\n",
    "                                  [555463, 555613],\n",
    "                                  [555798, 555871],\n",
    "                                  [560308, 560427],\n",
    "                                  [560520, 560754],\n",
    "                                  [561018, 561760],\n",
    "                                  [567871, 569759],\n",
    "                                  [574985, 575100],\n",
    "                                  [575173, 575901],\n",
    "                                  [576129, 576752],\n",
    "                                  [576905, 577403],\n",
    "                                  [581610, 581787],\n",
    "                                  [582457, 582533],\n",
    "                                  [598178, 598472],\n",
    "                                  [605392, 605481],\n",
    "                                  [605542, 605688],\n",
    "                                  [606044, 606165],\n",
    "                                  [606316, 606771],\n",
    "                                  [627113, 627250],\n",
    "                                  [629048, 629203],\n",
    "                                  [629475, 629605],\n",
    "                                  [632494, 632623],\n",
    "                                  [636919, 637208],\n",
    "                                  [637264, 637678],\n",
    "                                  [637755, 637924],\n",
    "                                  [638135, 638330],\n",
    "                                  [638676, 638986],\n",
    "                                  [639089, 640680],\n",
    "                                  [640795, 641207],\n",
    "                                  [641269, 641396],\n",
    "                                  [641474, 641653],\n",
    "                                  [642684, 642828],\n",
    "                                  [645639, 645785],\n",
    "                                  [645862, 646010],\n",
    "                                  [657719, 657805],\n",
    "                                  [659373, 659473],\n",
    "                                  [659735, 659808],\n",
    "                                  [661062, 661147],\n",
    "                                  [665413, 666282],\n",
    "                                  [678746, 678888],\n",
    "                                  [679341, 679713],\n",
    "                                  [680344, 680433],\n",
    "                                  [688882, 689014],\n",
    "                                  [689785, 689892],\n",
    "                                  [690406, 690524],\n",
    "                                  [690699, 692591],\n",
    "                                  [692884, 693088],\n",
    "                                  [693218, 696855],\n",
    "                                  [696927, 697179],\n",
    "                                  [697237, 697367],\n",
    "                                  [697421, 697706],\n",
    "                                  [697758, 698183],\n",
    "                                  [698274, 699233],\n",
    "                                  [699788, 699891],\n",
    "                                  [700003, 700187],\n",
    "                                  [702014, 702459],\n",
    "                                  [702546, 703740],\n",
    "                                  [705028, 707706],\n",
    "                                  [719666, 719806],\n",
    "                                  [720652, 721647],\n",
    "                                  [721712, 721834],\n",
    "                                  [721910, 722600],\n",
    "                                  [725511, 726402],\n",
    "                                  [726575, 726706],\n",
    "                                  [726779, 726910],\n",
    "                                  [727737, 727842],\n",
    "                                  [727923, 728169],\n",
    "                                  [728500, 728700],\n",
    "                                  [729341, 729446],\n",
    "                                  [737645, 737857],\n",
    "                                  [746434, 748304],\n",
    "                                  [748377, 749086],\n",
    "                                  [759919, 760564],\n",
    "                                  [761595, 763202],\n",
    "                                  [763340, 765869],\n",
    "                                  [765933, 766568],\n",
    "                                  [766973, 767132],\n",
    "                                  [767386, 767488],\n",
    "                                  [771719, 772165],\n",
    "                                  [772435, 773741],\n",
    "                                  [773793, 773875],\n",
    "                                  [776484, 776561],\n",
    "                                  [776621, 778465],\n",
    "                                  [779200, 779289],\n",
    "                                  [779593, 779700],\n",
    "                                  [779898, 779972],\n",
    "                                  [781657, 782002],\n",
    "                                  [782179, 783313],\n",
    "                                  [783464, 783735],\n",
    "                                  [784036, 784209],\n",
    "                                  [784420, 784783],\n",
    "                                  [784879, 785411],\n",
    "                                  [785742, 787327],\n",
    "                                  [787389, 787514],\n",
    "                                  [787577, 787753],\n",
    "                                  [787994, 788318],\n",
    "                                  [788617, 790287],\n",
    "                                  [790364, 790599],\n",
    "                                  [790769, 790949],\n",
    "                                  [791427, 791559],\n",
    "                                  [792918, 798735],\n",
    "                                  [799113, 799975],\n",
    "                                  [800045, 800345],\n",
    "                                  [800428, 800636],\n",
    "                                  [800702, 800861],\n",
    "                                  [800989, 801282],\n",
    "                                  [801369, 801442],\n",
    "                                  [801906, 802062],\n",
    "                                  [802731, 802806],\n",
    "                                  [802860, 802936],\n",
    "                                  [803021, 803150],\n",
    "                                  [803256, 803366],\n",
    "                                  [803472, 803707],\n",
    "                                  [803895, 805844],\n",
    "                                  [808638, 810571],\n",
    "                                  [818389, 820786],\n",
    "                                  [820880, 820998],\n",
    "                                  [821096, 821474],\n",
    "                                  [821609, 821811],\n",
    "                                  [822023, 822278],\n",
    "                                  [832265, 833924],\n",
    "                                  [834364, 835366]]\n",
    "                            }\n",
    "    chromosome = 'chr1'\n",
    "    start = 540470\n",
    "    stop = 540670\n",
    "    assert(are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[0]==False)\n",
    "    \n",
    "    chromosome = 'chr1'\n",
    "    start = 540510\n",
    "    stop = 540710\n",
    "    assert(are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[0]==False)\n",
    "    \n",
    "    import time\n",
    "    origTime = time.time()\n",
    "    chromosome = 'chr11'\n",
    "    start = 834310\n",
    "    stop = 834510\n",
    "    print time.time() - origTime\n",
    "    print are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[1]\n",
    "    assert(are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[0])\n",
    "    \n",
    "    origTime = time.time()\n",
    "    chromosome = 'chr11'\n",
    "    start = 187800\n",
    "    stop = 188000\n",
    "    assert(are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[0]==False)\n",
    "    print time.time()-origTime\n",
    "    \n",
    "    chromosome = 'chr11'\n",
    "    start = 187500\n",
    "    stop = 835368\n",
    "    print are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[1]\n",
    "    assert(are_coordinates_half_in_peaks(sorted_peak_coordinates, chromosome, start, stop)[0]==False)\n",
    "    \n",
    "    print sum([val[1]-val[0] for val in sorted_peak_coordinates['chr11']])\n",
    "    \n",
    "test_are_coordinates_half_in_peaks()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
